Fix issue with ScalateUrlGeneratorSupport and reverse-routes.

j1mr10rd4n opened the pull request:
This PR fixes #761.

When reverse routes are reflected on and saved to the scalate context, they now include information about the servlet in which they were declared.

Note that in order to avoid collisions, the variable names for reverse routes should be unique across all servlets. A warning to this effect was added to the User Guide in [scalatra-website PR 174](https://github.com/scalatra/scalatra-website/pull/174)

j1mr10rd4n commented:
Looks like the scala-2.12 builds have timed out - I saw this a couple of times locally also - is this a known issue since the move to sbt-1.0 ?

takezoe commented:
Travis build is unstable recently... I will rerun failed builds.

takezoe commented on scalate/src/main/scala/org/scalatra/scalate/ScalateUrlGeneratorSupport.scala:
```diff
@@ -18,16 +18,33 @@ trait ScalateUrlGeneratorSupport extends ScalateSupport {
+        val routeWithOwnServletPath = route.copy(contextPath = pathReplacingFn)
+        context.attributes.update(name, routeWithOwnServletPath)
+      }
+    }
```
Isn't it possible to use methods in `UrlGenerator` to assemble url?

j1mr10rd4n replied:
[UrlGenerator](https://github.com/scalatra/scalatra/blob/v2.6.2/core/src/main/scala/org/scalatra/UrlGenerator.scala) does assemble urls from [Routes](https://github.com/scalatra/scalatra/blob/v2.6.2/core/src/main/scala/org/scalatra/Route.scala) - the intent of the fix made to [ScalateUrlGeneratorSupport](https://github.com/scalatra/scalatra/blob/v2.6.2/scalate/src/main/scala/org/scalatra/scalate/ScalateUrlGeneratorSupport.scala) is to save Routes with the correct context paths to generate the urls from.

takezoe commented on core/src/main/scala/org/scalatra/servlet/RichServletContext.scala:
```diff
@@ -222,5 +222,17 @@ case class RichServletContext(sc: ServletContext) extends AttributesMap {
+    Option(sc.getAttribute(generatedUrlsKey))
+      .getOrElse(Map.empty[String, Map[String, Route]])
+      .asInstanceOf[Map[String, Map[String, Route]]]
+  }
```
I think that these code should be in `ScalateUrlGeneratorSupport` side because `RichServletContext` shouldn't know something about `ScalateUrlGeneratorSupport`.

j1mr10rd4n replied:
Good point!

takezoe commented on core/src/main/scala/org/scalatra/servlet/RichServletContext.scala:
```diff
@@ -222,5 +222,17 @@ case class RichServletContext(sc: ServletContext) extends AttributesMap {
+  val generatedUrlsKey = "org.scalatra.scalate.generatedUrls"
+
+  def addGeneratedUrls(servletClassName: String, routes: Map[String, Route]) = {
+    sc.setAttribute(generatedUrlsKey, generatedUrls + (servletClassName -> routes))
```
It would be nice if it warns name collision as you mentioned in the gitter room.

j1mr10rd4n replied:
Ok, I can add in a Logger instance and a warning.

j1mr10rd4n commented:
Hi @takezoe - I've done a bit of refactoring and moved all changes in to [ScalateUrlGeneratorSupport](https://github.com/scalatra/scalatra/blob/v2.6.2/scalate/src/main/scala/org/scalatra/scalate/ScalateUrlGeneratorSupport.scala) and added a log message at warning level if duplicate reverse routes are found.

takezoe commented:
@j1mr10rd4n By the way, reverse routing seems to be available only within a controller, originally because it requires a method reference.
http://scalatra.org//guides/2.6/http/reverse-routes.html

Unlike it, this pull request makes possible to get routes of all controllers. This difference might confuse users. Sorry we had to discuss about this difference in the first place but I noticed it after merging this pull request.

I would like to hear your opinion.
